rust/template/tcp_channel/lib.rs

```diff
@@ -0,0 +1,262 @@
+use std::net::TcpStream;
+use std::net::ToSocketAddrs;
+use std::ops::DerefMut;
+use std::os::unix::io::AsRawFd;
+use std::os::unix::io::RawFd;
+use std::sync::Arc;
+use std::sync::Mutex;
+use std::thread::spawn;
+use std::thread::JoinHandle;
+
+use libc::close;
+
+use observe::Observable;
+use observe::Observer;
+use observe::Subscription;
+use observe::UpdatesSubscription;
+
+use serde::de::DeserializeOwned;
+use serde::ser::Serialize;
+use serde_json::from_str;
+use serde_json::to_string;
+
+/// The receiving end of a TCP channel has an address
+/// and streams data to an observer.
+#[derive(Debug)]
+pub struct TcpReceiver<T> {
+    addr: SocketAddr,
+    listener: RawFd,
+    stream: Arc<Mutex<Option<RawFd>>>,
+    thread: JoinHandle<io::Result<()>>,
+    observer: Arc<Mutex<Option<Box<dyn Observer<T, String> + Send>>>>,
+}
+
+impl<T: DeserializeOwned + Send + Debug + 'static> TcpReceiver<T> {
+    /// Create a new TCP receiver with no observer.
+    ///
+    /// `addr` may have a port set (by setting it to 0). In such a case
+    /// the system will assign a port that is free. To retrieve this
+    /// assigned port (in the form of the full `SocketAddr`), use the
+    /// `addr` method.
+    pub fn new<A>(addr: A) -> io::Result<Self>
+    where
+        A: ToSocketAddrs,
+    {
+        let listener = TcpListener::bind(addr)?;
+        // We want to allow for auto-assigned ports, by letting the user
+        // specify a `SocketAddr` with port 0. In this case, after
+        // actually binding to an address, we need to update the port we
+        // got assigned in `addr`, but for simplicity we just copy the
+        // entire thing.
+        let addr = listener.local_addr()?;
+        let listener_fd = listener.as_raw_fd();
+        let stream_fd = Arc::new(Mutex::new(None));
+        let observer = Arc::new(Mutex::new(None));
+        let thread = Self::accept(stream_fd.clone(), listener, observer.clone());
+
+        Ok(Self {
+            addr,
+            listener: listener_fd,
+            stream: stream_fd,
+            thread,
+            observer,
+        })
+    }
+
+    /// Accept a connection (in a non-blocking manner), read data from
+    /// it, and dispatch that to the subscribed observer, if any. If no
+    /// observer is subscribed, data will be silently dropped.
+    fn accept(
+        stream: Arc<Mutex<Option<RawFd>>>,
+        listener: TcpListener,
+        observer: Arc<Mutex<Option<Box<dyn Observer<T, String> + Send>>>>,
+    ) -> JoinHandle<io::Result<()>> {
+        spawn(move || {
+            let (socket, _) = listener.accept()?;
+            *stream.lock().unwrap() = Some(socket.as_raw_fd());
+
+            let reader = BufReader::new(socket);
+            let mut lines = reader.lines();
+            loop {
+                let mut upds = lines
+                    .by_ref()
+                    .take_while(|result| match result {
+                        Ok(line) if line == "commit" => false,
```

Comment:
Should there be a "start" command as well or does "commit" implicitly start the next set of updates?

Reply:
"commit" acts as the marker that we have read enough and an indication that we can now actually start processing. There is no explicit "start" send because "commit" basically subsumes that.
This will change once we switch to using a binary encoding, in which case we will have a 1:1 relationship between events (`on_***` method calls) and messages being sent over the wire, so I am not going to change anything here for this patch.